### PR TITLE
[feature](function) Support SHA family functions

### PR DESCRIPTION
**Pull request by zclllyybb**
## Proposed changes

Issue Number: close #xxx

```SQL
mysql> select sha("123");
 +------------------------------------------+
 | sha1('123')                              |
 +------------------------------------------+
 | 40bd001563085fc35165329ea1ff5c5ecbdbbeef |
 +------------------------------------------+
 1 row in set (0.13 sec)

mysql> select sha2('abc', 224);
+----------------------------------------------------------+
| sha2('abc', 224)                                         |
+----------------------------------------------------------+
| 23097d223405d8228642a477bda255b32aadbce4bda0b3f7e36c9da7 |
+----------------------------------------------------------+
1 row in set (0.13 sec)

mysql> select sha2('abc', 384);
+--------------------------------------------------------------------------------------------------+
| sha2('abc', 384)                                                                                 |
+--------------------------------------------------------------------------------------------------+
| cb00753f45a35e8bb5a03d699ac65007272c32ab0eded1631a8b605a43ff5bed8086072ba1e7cc2358baeca134c825a7 |
+--------------------------------------------------------------------------------------------------+
1 row in set (0.13 sec)

mysql> select sha2(NULL, 512);
+-----------------+
| sha2(NULL, 512) |
+-----------------+
| NULL            |
+-----------------+
1 row in set (0.09 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: gensrc/script/doris_builtins_functions.py [hunk 1]
```diff
@@ -1901,6 +1901,9 @@
         [['sm3sum'], 'VARCHAR', ['VARCHAR', '...'], ''],
         [['sm3'], 'VARCHAR', ['STRING'], ''],
         [['sm3sum'], 'VARCHAR', ['STRING', '...'], ''],
+
+        [['sha2'], 'VARCHAR', ['VARCHAR', 'INT'], ''],
+        [['sha2'], 'VARCHAR', ['STRING', 'INT'], ''],
```
Comment (HappenLee) [inline gensrc/script/doris_builtins_functions.py:1906]: Add nereids function

File: be/src/vec/functions/function_string.h [hunk 2]
```diff
@@ -2045,6 +2045,107 @@ class FunctionStringMd5AndSM3 : public IFunction {
     }
 };
 
+class FunctionStringDigestSHA2 : public IFunction {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:2048]: Impl the SHA1 also

File: be/src/vec/functions/function_string.h [hunk 3]
```diff
@@ -2045,6 +2045,107 @@ class FunctionStringMd5AndSM3 : public IFunction {
...
+
+        block.replace_by_position(result, std::move(res_col));
+        return Status::OK();
+    }
+
+private:
+    void execute_224(const ColumnString::Chars& data, const ColumnString::Offsets& offset,
+                     int input_rows_count, ColumnString::Chars& res_data,
+                     ColumnString::Offsets& res_offset) {
+        SHA224Digest digest;
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:2098]: which can cause error

File: be/src/util/sha.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+void SHA512Digest::update(const void* data, size_t length) {
+    SHA512_Update(&_sha512_ctx, data, length);
+}
+
+void SHA512Digest::digest() {
+    unsigned char hash[SHA512_DIGEST_LENGTH];
+    SHA512_Final(hash, &_sha512_ctx);
+
+    std::stringstream ss;
```
Comment (HappenLee) [inline be/src/util/sha.cpp:98]: which cause the another cost

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/util/sha.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+std::string_view SHA1Digest::digest() {
+    unsigned char buf[SHA_DIGEST_LENGTH];
+    SHA1_Final(buf, &_sha_ctx);
+
+    char* to = _reuse_hex;
+    for (int i = 0; i < SHA_DIGEST_LENGTH; ++i) {
+        *to++ = dig_vec_lower[buf[i] >> 4];
+        *to++ = dig_vec_lower[buf[i] & 0x0F];
```
Comment (BiteTheDDDDt) [inline be/src/util/sha.cpp:40]: Define constant as constexpr variable

File: be/src/vec/functions/function_string.h [hunk 6]
```diff
@@ -2045,6 +2045,107 @@ class FunctionStringMd5AndSM3 : public IFunction {
...
+    size_t get_number_of_arguments() const override { return 1; }
+    bool is_variadic() const override { return true; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeString>();
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) override {
+        DCHECK(arguments.size() == 1);
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_string.h:2062]: use DCHECK_EQ

**Comment by zclllyybb** [comment]
run buildall

**Review by BiteTheDDDDt — DISMISSED** [review]
LGTM

**Comment by zclllyybb** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**